### PR TITLE
crypto: update BLS signature and public key serialization

### PR DESCRIPTION
**Pull request by tarakby**
This PR updates the serialization of G1 points (signatures) and G2 points (public keys) to be compliant with the [IETF draft](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-08.html#name-zcash-serialization-format-) of pairing-friendly curves.

Since the updates require recent Relic tools, the Relic submodule is updated.

- update Relic submodule version
- Integrate the recent changes in Relic (Jacobian coordinates, renaming)
- fix an issue with the optimized SWU hashing to curve when point addition requires the coefficient `a`.
- update the serialization of G1 points using the "sign" of coordinate `y` instead of parity.
- update the serialization of G2 points using the "sign" of coordinate `y` instead of parity.

### CODE REVIEW

File: crypto/bls12381_hashtocurve.c [hunk 1]
```diff
@@ -215,7 +73,8 @@ const uint64_t p_3div4_data[Fp_DIGITS] = {
     0xD91DD2E13CE144AF, 0x92C6E9ED90D2EB35, 0x0680447A8E5FF9A6,
 };
 
-const uint64_t p_1div2_data[Fp_DIGITS] = {
+// convertede to Montgomery form
```
Comment (tarakby) [inline crypto/bls12381_hashtocurve.c:76]: typo

File: crypto/bls12381_hashtocurve.c [hunk 2]
```diff
@@ -572,22 +431,28 @@ static void map_to_G1_opswu(ep_t p, const uint8_t *msg, int len) {
...
+        eval_iso11(p_temp, p_temp); // map to E
+        // second mapping
         map_to_E1_swu(p, t2); // map to E1
-        ep_add_projc(p, p, p_temp);
-        eval_iso11(p_temp, p); // map to E
+        eval_iso11(p, p); // map to E
+        // sum and clear the cofactor
+        // TODO: implement point addition in E1 and apply the isogeny map
+        // only once.
+        ep_add_jacob(p, p, p_temp);
```
Comment (tarakby) [inline crypto/bls12381_hashtocurve.c:443]: add linearity comment

File: crypto/bls12381_utils.c [hunk 3]
```diff
@@ -38,6 +38,12 @@ extern const uint64_t beta_data[Fp_DIGITS];
 extern const uint64_t z2_1_by3_data[2];
 #endif
 
+const uint64_t p_1div2_data[Fp_DIGITS] = {
```
Comment (tarakby) [inline crypto/bls12381_utils.c:41]: precomputed p-1/2

File: crypto/bls12381_utils.c [hunk 4]
```diff
@@ -217,124 +219,49 @@ void bn_map_to_Zr_star(bn_t a, const uint8_t* bin, int len) {
...
-    bn_new(tmp);
-    fp_prime_back(tmp, a);
-    int res = bn_get_bit(tmp, bit);
-    bn_free(tmp);
-    return res;
-#else
-    return fp_get_bit(a, bit);
-#endif
+// returns the sign of y.
+// 1 if y > (p - 1) and 0 otherwise.
```
Comment (tarakby) [inline crypto/bls12381_utils.c:223]: 1 if y > (p-1)/2

File: crypto/bls12381_utils.c [hunk 5]
```diff
@@ -391,106 +318,64 @@ int ep_read_bin_compact(ep_t a, const byte *bin, const int len) {
...
 
-    TRY {
+    RLC_TRY {
         ep2_new(t);
         ep2_norm(t, (ep2_st *)a);
         fp2_write_bin(bin, 2*Fp_BYTES, t->x, 0);
 
         if (SERIALIZATION == COMPRESSED) {
-            bin[0] |= (fp_get_bit_generic(t->y[0], 0) << 5);
+            bin[0] |= (fp2_get_sign(t->y) << 5);
```
Comment (tarakby) [inline crypto/bls12381_utils.c:365]: look again

File: crypto/sign_test_utils.go [hunk 6]
```diff
@@ -22,14 +21,14 @@ func testGenSignVerify(t *testing.T, salg SigningAlgorithm, halg hash.Hasher) {
 	input := make([]byte, 100)
 	mrand.Seed(time.Now().UnixNano())
 
-	loops := 50
+	loops := 3
```
Comment (tarakby) [inline crypto/sign_test_utils.go:24]: 50

File: crypto/sign_test_utils.go [hunk 7]
```diff
@@ -201,19 +207,20 @@ func benchSign(b *testing.B, algo SigningAlgorithm, halg hash.Hasher) {
 // testPOP tests proofs of possession
 func testPOP(t *testing.T, salg SigningAlgorithm, halg hash.Hasher) {
 	t.Logf("Testing proof of possession for %s", salg)
+	mrand.Seed(time.Now().UnixNano())
```
Comment (tarakby) [inline crypto/sign_test_utils.go:210]: to print

Comment (tarakby) [reply crypto/sign_test_utils.go:210]: t.Logf("%d", seed)

**Review by ramtinms — APPROVED** [review]
👏

**Review by jordanschalm — APPROVED** [review]
Tarak walked us through on a call

**Review by zhangchiqing — APPROVED** [review]

File: crypto/sign_test_utils.go [hunk 8]
```diff
@@ -22,14 +21,14 @@ func testGenSignVerify(t *testing.T, salg SigningAlgorithm, halg hash.Hasher) {
 	input := make([]byte, 100)
 	mrand.Seed(time.Now().UnixNano())
 
-	loops := 50
+	loops := 3
 	for j := 0; j < loops; j++ {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/sign_test_utils.go:25]: 1. Yep, that's testing KeyGen is deterministic. It is not a property of a signature scheme, but a property we may want the implementation to have, like ours. That test is actually covered [here](https://github.com/onflow/flow-go/blob/tarak/5025-general-updates/crypto/sign_test_utils.go#L135): if we use the same seed, then the keys are equal.  The comparison is done using the math structures, which guarantees the signing and verifying are consistent. 

2. That test is missing, I'll add it 👌

File: crypto/sign_test_utils.go [hunk 9]
```diff
@@ -69,56 +68,63 @@ func testGenSignVerify(t *testing.T, salg SigningAlgorithm, halg hash.Hasher) {
...
 
-	// Key generation seed
-	seed := make([]byte, seedMinLength)
-	read, err := rand.Read(seed)
-	require.Equal(t, read, seedMinLength)
-	require.NoError(t, err)
-	sk, err := GeneratePrivateKey(salg, seed)
-	assert.Nil(t, err, "the key generation has failed")
+	loops := 50
+	for j := 0; j < loops; j++ {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/sign_test_utils.go:76]: Good suggestion, I'll add that too.

File: crypto/sign_test_utils.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/sign_test_utils.go:27]: addressed [here](https://github.com/onflow/flow-go/pull/117/commits/c8e119a59bcb914bca9cb77ec74b216da6c4f1fa)

File: crypto/sign_test_utils.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/sign_test_utils.go:84]: addressed (here)[https://github.com/onflow/flow-go/pull/117/commits/f71ceb442e711e000e93acc8ac15fa08365d204b]

**Pull request merged**